### PR TITLE
fix eni lookup logic to update them only when univocally specified

### LINKED ISSUE #19972: ec2_eni doesn't honour description for uniqueness
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
ec2_eni

##### ANSIBLE VERSION
```
ansible 2.2.0.0
  config file = /Users/vlfig/projects/ansible-confluent-orchestration/ansible.cfg
  configured module search path = Default w/o overrides
```

Reproducible in 
```
ansible 2.3.0 (devel ad1e066540) last updated 2017/01/06 11:47:13 (GMT +100)
  config file = /Users/vlfig/projects/ansible-confluent-orchestration/ansible.cfg
  configured module search path = Default w/o overrides
```

*Not* reproducible in 2.1.3.0 (works as expected).

##### CONFIGURATION
plain

##### OS / ENVIRONMENT
N/A

##### SUMMARY
The ec2_eni module does not create a new ENI and instead updates the description of an existing ENI that matches its other args. This is important for the cases when we don't know the IP address nor want to attach it to an instance and just want to create an ENI.

##### STEPS TO REPRODUCE

```yaml
- ec2_eni:
    description: "my-eni-1"
    region: "{{ region }}"
    subnet_id: "{{ my_subnet }}"
    state: present
    security_groups: "{{ my_sg }}"
```
and then
```yaml
- ec2_eni:
    description: "my-eni-2"
    region: "{{ region }}"
    subnet_id: "{{ my_subnet }}"
    state: present
    security_groups: "{{ my_sg }}"
```

(one could also use `with_items` with the same results)

##### EXPECTED RESULTS

Two ENIs created with descriptions "my-eni-1" and "my-eni-2".

##### ACTUAL RESULTS

Only one ENI gets created, with description "my-eni-2". First task creates, second task updates the same ENI.

Note `"id": "eni-57cf9d28"` is the same in both invocations.
```
 _____________________________
< TASK [test-image : ec2_eni] >
 -----------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

task path: [snip]/roles/test-image/tasks/main.yml:34
Using module file [snip]/lib/python2.7/site-packages/ansible/modules/extras/cloud/amazon/ec2_eni.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: vlfig
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/env python && sleep 0'
changed: [localhost] => {
    "changed": true,
    "interface": {
        "description": "my-eni-1",
        "groups": {
            "sg-483a9d20": "[snip]"
        },
        "id": "eni-57cf9d28",
        "mac_address": "06:2f:d2:cb:38:f7",
        "owner_id": "[snip]",
        "private_ip_address": "10.1.12.57",
        "private_ip_addresses": [
            {
                "primary_address": true,
                "private_ip_address": "10.1.12.57"
            }
        ],
        "source_dest_check": true,
        "status": "pending",
        "subnet_id": "subnet-1f3b6364",
        "vpc_id": "vpc-[snip]cb"
    },
    "invocation": {
        "module_args": {
            "attached": null,
            "aws_access_key": "[snip]",
            "aws_secret_key": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "delete_on_termination": null,
            "description": "my-eni-1",
            "device_index": 0,
            "ec2_url": null,
            "eni_id": null,
            "force_detach": false,
            "instance_id": null,
            "private_ip_address": null,
            "profile": null,
            "region": "eu-central-1",
            "secondary_private_ip_address_count": null,
            "secondary_private_ip_addresses": null,
            "security_groups": [
                "sg-483a9d20",
                "sg-483a9d20"
            ],
            "security_token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "source_dest_check": null,
            "state": "present",
            "subnet_id": "subnet-1f3b6364",
            "validate_certs": true
        },
        "module_name": "ec2_eni"
    }
}
 _____________________________
< TASK [test-image : ec2_eni] >
 -----------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

task path: [snip]/roles/test-image/tasks/main.yml:44
Using module file [snip]/lib/python2.7/site-packages/ansible/modules/extras/cloud/amazon/ec2_eni.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: vlfig
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/env python && sleep 0'
changed: [localhost] => {
    "changed": true,
    "interface": {
        "description": "my-eni-2",
        "groups": {
            "sg-483a9d20": "[snip]"
        },
        "id": "eni-57cf9d28",
        "mac_address": "06:2f:d2:cb:38:f7",
        "owner_id": "[snip]",
        "private_ip_address": "10.1.12.57",
        "private_ip_addresses": [
            {
                "primary_address": true,
                "private_ip_address": "10.1.12.57"
            }
        ],
        "source_dest_check": true,
        "status": "available",
        "subnet_id": "subnet-1f3b6364",
        "vpc_id": "vpc-[snip]cb"
    },
    "invocation": {
        "module_args": {
            "attached": null,
            "aws_access_key": "[snip]",
            "aws_secret_key": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "delete_on_termination": null,
            "description": "my-eni-2",
            "device_index": 0,
            "ec2_url": null,
            "eni_id": null,
            "force_detach": false,
            "instance_id": null,
            "private_ip_address": null,
            "profile": null,
            "region": "eu-central-1",
            "secondary_private_ip_address_count": null,
            "secondary_private_ip_addresses": null,
            "security_groups": [
                "sg-483a9d20",
                "sg-483a9d20"
            ],
            "security_token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "source_dest_check": null,
            "state": "present",
            "subnet_id": "subnet-1f3b6364",
            "validate_certs": true
        },
        "module_name": "ec2_eni"
    }
}
[snip]
localhost                  : ok=12   changed=3    unreachable=0    failed=0
```

### PR DESCRIPTION
**Pull request by vlfig**
(Fixes #19972)

Also remove redundant eni lookup.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
ec2_eni

##### ANSIBLE VERSION
```
ansible 2.2.0.0
  config file = /Users/vlfig/projects/ansible-confluent-orchestration/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY
When no eni_id, private_ip_address nor (instance_id and device_index) are provided this module should not update existing ENI.

This change makes it so that only when parameters that actually univocally identify an ENI are provided the existing ENI (if any) is updated. Otherwise a new one gets created.

### CODE REVIEW

**Comment by zyv** [comment]
This fixes the problem for me, indeed!

I wish, however, that `ec2_eni` supported tags like many other `ec2_*` modules to uniquely identify items, because right now this means that one would have to hardcode an IP address / `eni_id` to make provisioning runs idempotent.

This is a bit annoying given that the `subnet_id` argument is mandatory, and so one can't just specify a private IP address and have subnet deduced on this basis, but rather hardcode both at the same time...

**Comment by wimnat** [comment]
Private IP is not unique. You could choose the same IP across multiple
VPCs.

Subnet ID is easy to look up using tags or any other filter using
ec2_eni_facts module.

**Comment by zyv** [comment]
On Mon, 16 Jan 2017, Rob wrote:

> Private IP is not unique. You could choose the same IP across multiple 
> VPCs.

True, which is why it would be awesome to be able to use tags for 
uniqueness with this module, like with `ec2_vpc`, for example, or others.

Don't get me wrong, I'm not saying this to block the pull request, on the 
contrary, I've reported that I've tested it and it fixes the original 
problem for me.

> Subnet ID is easy to look up using tags or any other filter using 
> ec2_eni_facts module.

Yes, that's what I'll end up using for now, I think.

-- 
Sincerely yours,
Yury V. Zaytsev

**Comment by zyv** [comment]
@wimnat @vlfig I have tested the suggested solution today:

```
- name: 'gather facts about {{ aws_vpc_name }} public subnets'

  ec2_vpc_subnet_facts:
    region: '{{ aws_region }}'
    filters:
      'vpc-id': '{{ aws_vpc_id }}'
      'tag:Scope': 'public'

  register: result_aws_vpc_subnets_public

- name: create vpn instance network interface

  ec2_eni:
    region: '{{ aws_region }}'
    subnet_id: '{{ item.id }}'
    description: 'OpenVPN server network interface'
    security_groups: ['ansible-{{ aws_vpc_name }}-vpn-server']
    source_dest_check: False
    private_ip_address: '{{ item.cidr | regex_replace("\.([0-9]+/[0-9]+)$", ".254") }}'
    state: present

  with_items: '{{ result_aws_vpc_subnets_public | json_query("subnets[*].{id: id, cidr: cidr_block}") }}'

  register: result_aws_vpc_vpn_eni
```

After I run this playbook for the second time I get the following error:

```
TASK [create vpn instance network interface] **************
failed: [localhost] (item={'cidr': u'1.2.3.0/20', 'id': u'subnet-XXX'}) => {"failed": true, "item": {"cidr": "1.2.3.0/20", "id": "subnet-XXX"}, "msg": "InterfaceAddressModel(interfaceId=interface-UUU, address=1.2.3.254, subnetId=subnet-XXX, addressIndex=0) is already assigned."}
failed: [localhost] (item={'cidr': u'1.2.4.0/20', 'id': u'subnet-YYY'}) => {"failed": true, "item": {"cidr": "1.2.4.0/20", "id": "subnet-YYY"}, "msg": "InterfaceAddressModel(interfaceId=interface-VVV, address=1.2.4.254, subnetId=subnet-YYY, addressIndex=0) is already assigned."}
```

I expected the module to return OK instead and an idempotent run to continue :-(

**Comment by zyv** [comment]
FYI, I have tried with the module shipped with current stable `ansible-2.2.1.0` and there the playbook above seems to work, and is idempotent. So it appears that the pull request is introducing this problem.

**Comment by willthames** [comment]
needs_revision

Need to address the issues @zyv raises

File: lib/ansible/modules/cloud/amazon/ec2_eni.py [hunk 1]
```diff
@@ -458,28 +456,32 @@ def detach_eni(eni, module):
         module.exit_json(changed=False, interface=get_eni_info(eni))
 
 
-def find_eni(connection, module):
+def univocally_find_eni(connection, module):
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_eni.py:459]: I suspect 'univocally' should be 'uniquely' or 'unequivocally' (uniquely is likely simpler!)

**Comment by ryansb** [comment]
Merged as #23404

**Pull request closed without merging**